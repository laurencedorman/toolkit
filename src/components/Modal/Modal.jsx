// @flow
import React, { Component } from 'react';
import {
  Transition, animated, interpolate, config,
} from 'react-spring';

import { Icon, Portal } from 'components';
import styles from './Modal.module.scss';

/**
 * @visibleName Modal
 */
type propTypes = {
  on: boolean,
  toggle: () => void,
  children: string | Node,
};
/* eslint-disable */
export default class Modal extends Component<propTypes>{
  componentDidMount() {
    document.addEventListener('keydown', this.handleKey, false);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKey, false);
  }

  handleKey = e => {
    const { on, toggle } = this.props;

    const keys = {
      Escape: () => {
        e.preventDefault();
        !on
          ? e.stopPropagation()
          : toggle();
      },
    };

    keys[e.key]
      && keys[e.key]()
  };

  render() {
    const { on, toggle, children } = this.props;

    return (
      <Portal>
        <Transition
          native
          config={config.stiff}
          items={on}
          from={{ o: 0.01, s: 0.6, y: '-200px' }}
          enter={{ o: 1, s: 1, y: '0px' }}
          leave={{ o: 0.01, s: 0.6, y: '200px' }}
        >
          {on => on
            && (
              ({ o, s, y}) => (
                <animated.div
                  onClick={toggle}
                  className={styles.modal}
                  role="Dialog"
                  style={{
                    opacity: o.interpolate(o => o)
                  }}
                >
                  <animated.div
                    className={styles.content}
                    onClick={e => e.stopPropagation()}
                    role="Contentinfo"
                    style={{
                      transform: interpolate(
                        [s, y],
                        (s, y) => `scale(${s}) translate3d(0, ${y}, 0)`
                      )
                    }}
                  >
                    <Icon
                      name="close-circle"
                      size="26"
                      onClick={toggle}
                      className={styles.icon}
                    />
                    <div>{children}</div>
                  </animated.div>
                </animated.div>
              )
            )
          }
        </Transition>
      </Portal>
    )
  }
}
/* eslint-enable */
