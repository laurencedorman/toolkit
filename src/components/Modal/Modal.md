


modal with props trigger :
```js
initialState = {
  show: false,
};

<div id="root">
    <button onClick={() => setState({ show: true })}>open by controlled props</button>
    <Modal
      open={state.show}
      close={() => setState({ show: false })}>
      <p>My money's in that office, right? If she start giving me some bullshit about it ain't there, and we got to go someplace else and get it, I'm gonna shoot you in the head then and there.</p>
    </Modal>
</div>
```
