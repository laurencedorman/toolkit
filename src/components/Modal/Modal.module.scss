@import 'styles/import.scss';

.modal {
  position: fixed;
  display: flex;
  top: 0; right: 0; bottom: 0; left: 0;
  justify-content: center;
  align-items: center;
  background-color: rgba($manoBlack, .5);
  cursor: pointer;
  z-index: z($layers, modal);

  .icon {
    position: absolute;
    right: .75rem;
    cursor: pointer;
    transition: hover .15s ease;
    cursor: pointer;
    z-index: z($levels, max);
    &:hover svg { fill: $hover }
  }

  .header {
    justify-content: center;
    & .icon { top: .75em }

    &.hasContent {
      border-bottom: 1px solid $greyMed;

      & .icon { top: initial }
    }
  }

  .footer {
    justify-content: flex-end;
    border-top: 1px solid $greyMed;
  }
}

.content {
  position: absolute;
  display: flex;
  flex-flow: column wrap;
  min-width: 320px;
  max-width: 50%;
  max-height: 90vh;
  background: $white;
  border-radius: $radius;
  border: $border transparent;
  box-sizing: border-box;
  text-align: left;
  cursor: default;
  overflow-y: auto;
  z-index: z($levels, max);

  &.noFooter { padding-bottom: 2em }
}
