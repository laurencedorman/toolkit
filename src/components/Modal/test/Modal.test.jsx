import React from 'react';
import { shallow } from 'enzyme';
import Modal from '../Modal';

describe('Modal', () => {
  const wrapper = shallow(
    <Modal><p>Modal test</p></Modal>,
  );

  it('should render without crash', () => {
    expect(wrapper.length).toEqual(1);
  });
});
