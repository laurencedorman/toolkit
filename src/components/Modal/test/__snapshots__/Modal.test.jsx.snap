// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
<Portal>
  <Transition
    config={
      Object {
        "friction": 23,
        "mass": 0.6,
        "tension": 350,
      }
    }
    enter={
      Object {
        "o": 1,
        "s": 1,
        "y": "0",
      }
    }
    from={
      Object {
        "o": 0,
        "s": 0.75,
        "y": "33vh",
      }
    }
    items={false}
    keys={[Function]}
    leave={
      Object {
        "o": 0,
        "s": 0.75,
        "y": "33vh",
      }
    }
    native={true}
    reset={false}
    unique={false}
  >
    <Component />
  </Transition>
</Portal>
`;
